sklearn/datasets/base.py

```diff
@@ -823,3 +824,134 @@ def _pkl_filepath(*args, **kwargs):
+
+    Returns
+    -------
+    md5hash : String
+        MD5 hash of the file at the provided path.
+
+    """
+
+    md5hash = hashlib.md5()
+    chunk_size = 8192
+    with open(path, "rb") as f:
+        while 1:
+            buffer = f.read(chunk_size)
+            if not buffer:
+                break
+            md5hash.update(buffer)
+    return md5hash.hexdigest()
+
+
+def _validate_file_md5(expected_checksum, path):
+    """Compare the MD5 checksum of a file at a path with
+    an expected MD5 checksum. If they do not match,
+    remove the file at path and throw a ValueError.
+
+    Parameters
+    -----------
+    expected_checksum: String
+        Expected MD5 checksum of file at path.
+
+    path: String
+        Path of file to compare MD5 hash of.
+
+    """
+
+    if expected_checksum != _md5(path):
+        # remove the corrupted file
+        remove(path)
+        raise IOError("{} has an MD5 hash differing "
+                      "from expected, file may be "
+                      "corrupted.".format(path))
+
+
+def _fetch_url(url, path, checksum):
+    """
+    Fetch a dataset from a URL and check the MD5 checksum to ensure
+    fetch was completed and the correct file was downloaded
+
+    Parameters
+    -----------
+    URL: String
+        URL to fetch the download from.
+
+    path: String
+        Path to save the file to.
+
+    checksum: String
+        MD5 checksum to verify against the data
+
+    """
+
+    resume_url_downloader = PartialURLOpener()
+    path_temp = path + ".part"
+    if exists(path_temp):
+        # since path_temp exists, resume download
+        temp_file = open(path_temp, "ab")
+        # get the amount of path_temp we've downloaded
+        existing_size = getsize(path_temp)
+        request_range = 'bytes={}-'.format(existing_size)
+
+        print("Resuming download from {}, "
+              "already have {} bytes".format(url, existing_size),
+              file=sys.stderr)
+        resume_url_downloader.addheader("Range", request_range)
+
+        try:
+            # Try to download only the remainder of the file
+            dataset_url = resume_url_downloader.open(url)
+            # get the content range of the request
+            content_range = dataset_url.info().get('Content-Range')
+            if (content_range is None or
+                    not content_range.startswith(request_range)):
+                raise IOError("Server does not support the HTTP Range "
+                              "header, cannot resume download.")
+        except Exception:
+            # delete the temp file and retry download of whole file
+            remove(path_temp)
+            print("Attempting to re-download file after {!r}.".format(exec),
+                  file=sys.stderr)
+            _fetch_url(url, path, checksum)
+    else:
+        # no path_temp, so download from scratch
+        temp_file = open(path_temp, "wb")
+        dataset_url = resume_url_downloader.open(url)
+    while 1:
```

Comment:
regarding @jnothman comment [here](https://github.com/scikit-learn/scikit-learn/pull/7429#discussion_r114064999), `with` can not be used since `temp_file` is opened in an `if-else` structure with different parameters. 

However we might be able to rewrite the whole function use `with` to open `dataset_url` then check if the path exist and then use two different `with`